kahypar/partition/bin_packing/prepacking.h

```diff
@@ -0,0 +1,215 @@
+    sum += w;
+    weights[i - 1] = {w, sum};
+  }
+  const BalanceSegTree seg_tree(weights, max_k);
+  std::vector<PartitionID> parts;
+
+  ASSERT(nodes.size() > 0);
+  ASSERT([&]() {
+    for (size_t i = 0; i < static_cast<size_t>(context.partition.k); ++i) {
+      if (upper_weight[i] < context.initial_partitioning.perfect_balance_partition_weight[i]
+        || upper_weight[i] >= num_bins_per_part[i] * max_bin_weight
+        ) {
+        return false;
+      }
+    }
+    return true;
+  } (), "The allowed partition or bin weights are to small to find a valid partition.");
+
+  size_t i;
+  std::pair<PartitionMapping, std::vector<HypernodeWeight>> packing_result = packer.applySecondLevel(upper_weight, num_bins_per_part);
+  for (i = 0; i < nodes.size(); ++i) {
+    ASSERT(upper_weight.size() == packing_result.second.size());
+
+    const size_t max_part_idx = getMaxPartIndex(context, packing_result.second, weights[i].first, true, max_bin_weight);
+    const HypernodeWeight remaining = std::max(0, std::min(packing_result.second[max_part_idx] - upper_weight[max_part_idx]
+                                      + weights[i].second, weights[i].second));
+
+    // find subrange of specified weight
+    const size_t j = weights.crend() - std::lower_bound(weights.crbegin(), weights.crend() - i, std::make_pair(0, remaining),
+                     [&](const auto& v1, const auto& v2) {
+                       return v1.second < v2.second;
+                     }) - 1;
+
+    // calculate the bound for the subrange
+    if (j > i) {
+      const HypernodeWeight imbalance = seg_tree.query(i, j - 1) + weights[j].second;
+      const HypernodeWeight partWeight = (j == nodes.size() ? packing_result.second[max_part_idx] + weights[i].second : upper_weight[max_part_idx])
+                                         * max_k / num_bins_per_part[max_part_idx];
+      if (partWeight + imbalance <= max_k * max_bin_weight) {
+        break;
+      }
+    }
+
+    parts.push_back(packer.insertElement(weights[i].first));
+    packing_result = packer.applySecondLevel(upper_weight, num_bins_per_part);
+  }
+  ASSERT(parts.size() <= nodes.size());
+
+  packing_result.first.applyMapping(parts);
+  for (size_t i = 0; i < parts.size(); ++i) {
+    hg.setFixedVertex(nodes[i], parts[i]);
+  }
+
+  // TODO(maas) It is questionable whether this last part of the algorithm is required.
+  // It is an optimization which theoretically allows to relax the upper allowed partition
+  // weights for the bisection in some cases. However, we also do know that more relaxed weights
+  // in a higher bisection level might lead to worse results in lower levels or in the refinement.
+  // Thus, in practice it is probably no improvement.
+  //
+  // The ugly aspect here is that it also complicates code at other positions, because the changed
+  // weights need to be propagated to the initial partitioning algorithm (see initial_partition::partition).
```

Comment:
Any thoughts on this?

Reply:
Could we do some experiments to check whether or not it is beneficial to use this?

Reply:
I have started an experiment comparing two variants where this is enabled/disabled. Hopefully, the results will show whether there is an impact on the result quality.

Reply:
The result showed a slightly negative impact on result quality, therefore I removed this part of the algorithm.